lib/ae_mdw/blocks.ex

```diff
@@ -0,0 +1,190 @@
+
+    case intersect_scopes([range_scope, cursor_scope, global_scope], direction) do
+      {:ok, first, last} when last - first > limit ->
+        {render_blocks(first, first + limit - 1, last_gen, expand?),
+         serialize_cursor(first + limit)}
+
+      {:ok, first, last} when first - last > limit ->
+        {render_blocks(first, first - limit + 1, last_gen, expand?),
+         serialize_cursor(first - limit)}
+
+      {:ok, first, last} ->
+        {render_blocks(first, last, last_gen, expand?), nil}
+
+      :empty ->
+        {[], nil}
+    end
+  end
+
+  defp intersect_scopes(scopes, direction) do
+    scopes
+    |> Enum.reject(&is_nil/1)
+    |> Enum.reduce(fn
+      {first, last}, {acc_first, acc_last} when direction == :forward ->
+        {max(first, acc_first), min(last, acc_last)}
+
+      {first, last}, {acc_first, acc_last} ->
+        {min(first, acc_first), max(last, acc_last)}
+    end)
+    |> case do
+      {first, last} when direction == :forward and first <= last -> {:ok, first, last}
+      {_first, _last} when direction == :forward -> :empty
+      {first, last} when direction == :backward and first >= last -> {:ok, first, last}
+      {_first, _last} when direction == :backward -> :empty
+    end
+  end
+
+  defp render_blocks(from, to, last_gen, expand?),
+    do: Enum.map(from..to, &render(&1, last_gen, expand?))
+
+  defp render(gen, last_gen, expand?) when gen > last_gen - @blocks_cache_threshold do
+    [key_block | micro_blocks] = fetch_gen_blocks(gen, last_gen)
+
+    put_mbs_from_db(key_block, micro_blocks, expand?)
+  end
+
+  defp render(gen, last_gen, expand?) do
+    [key_block | micro_blocks] = fetch_gen_blocks(gen, last_gen)
+
+    fetch_gen_from_cache(gen, key_block, micro_blocks, expand?)
+  end
+
+  defp fetch_gen_blocks(last_gen, last_gen) do
+    # gets by height once the chain current generation might happen to be higher than last_gen in DB
+    {:ok, %{key_block: kb, micro_blocks: mbs}} =
+      :aec_chain.get_generation_by_height(last_gen, :forward)
+
+    ^last_gen = :aec_blocks.height(kb)
+
+    for block <- [kb | mbs] do
+      header = :aec_blocks.to_header(block)
+      :aec_headers.serialize_for_client(header, Util.prev_block_type(header))
+    end
+  end
+
+  defp fetch_gen_blocks(gen, _last_gen) do
+    @table
+    |> Collection.stream(:backward, nil, {gen, <<>>})
+    |> Stream.take_while(&match?({^gen, _mb_index}, &1))
+    |> Enum.map(fn key -> Mnesia.fetch!(@table, key) end)
+    |> Enum.reverse()
+    |> Enum.map(fn block -> Format.to_map(block) end)
+  end
+
+  defp fetch_gen_from_cache(gen, key_block, micro_blocks, expand?) do
+    case EtsCache.get(@cache_table, gen) do
+      {key_block, _indx} ->
+        key_block
+
+      nil ->
+        key_block = put_mbs_from_db(key_block, micro_blocks, expand?)
+        EtsCache.put(@cache_table, gen, key_block)
+        key_block
+    end
+  end
+
+  defp put_mbs_from_db(key_block, micro_blocks, false) do
```

Comment:
This is where the piece of code should belong, the controller functions (incorrectly placed) are only used for integration tests to work. I don't like having additional code to maintain simply because of tests, the codebase should be as small as possible and tests should adapt to it, not the other way round. 

That said, what do you suggest we should do with `Cont`, `db_stream` and all of those functions exclusively used by integration tests? I suggest we get rid of them, since we have already tested that they match. Many of the integration tests will no longer be needed (I find testing pagination implementations against itself to be pretty useless). But from now on, every integration test we add should validate what that endpoint is supposed to return (e.g. when requesting `contract_call` txs, make sure that all of the results are contract calls), and not test the results against itself.

Thoughts?

Reply:
The suggestion is to use only this one from `AeMdw.Blocks` and get rid from those in `AeMdwWeb.BlockController`.

Reply:
Done, removed all of the Cont code